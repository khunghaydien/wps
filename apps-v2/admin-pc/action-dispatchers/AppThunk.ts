import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

// import { State } from '../modules';

export type AppAction<ReturnType = void> = ThunkAction<
  ReturnType,
  any,
  void,
  Action<string>
>;

export type AppDispatch = ThunkDispatch<unknown, void, AnyAction>;
