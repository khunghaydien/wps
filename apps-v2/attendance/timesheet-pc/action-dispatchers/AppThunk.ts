import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { State } from '../modules';

export type AppAction<ReturnType = unknown> = ThunkAction<
  ReturnType,
  State,
  void,
  Action<unknown>
>;

export type AppDispatch = ThunkDispatch<
  State,
  void,
  Action<unknown | AppAction>
>;
