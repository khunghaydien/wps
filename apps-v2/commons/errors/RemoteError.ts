import { Failure } from '../api/Response';

/**
 * Representing error occurred from remote
 */
export default class RemoteError extends Error {
  /**
   * Error code of API Error
   */
  errorCode: string;

  /**
   * Error message of API Error
   */
  message: string;

  /**
   * Stacktrace of API Error
   */
  stackTrace: string;

  /**
   * Error group
   *
   * 1: Operation error
   * 2: System error
   */
  groupCode: 1 | 2;

  /**
   * API path for Sentry Logging
   */
  path: string;

  constructor({ error }: Failure, path?: string) {
    super();

    this.errorCode = error.errorCode;
    this.message = error.message;
    this.stackTrace = error.stackTrace;
    this.stack = error.stackTrace;
    this.groupCode = error.groupCode;
    this.path = path || '';
  }
}
