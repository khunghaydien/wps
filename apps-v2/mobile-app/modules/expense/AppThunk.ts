import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

import { State } from './index';

export type AppDispatch = ThunkDispatch<
  { expense: State },
  void,
  Action<string>
>;
