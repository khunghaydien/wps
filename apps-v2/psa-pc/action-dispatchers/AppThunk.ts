import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { State } from '@psa/modules';

export type AppAction<ReturnType = void> = ThunkAction<
  ReturnType,
  State,
  void,
  Action<string>
>;

export type AppDispatch = ThunkDispatch<unknown, void, Action<string>>;
