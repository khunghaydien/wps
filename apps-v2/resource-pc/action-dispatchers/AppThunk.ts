import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { State } from '@resource/modules';

export type AppAction<ReturnType = void> = ThunkAction<
  ReturnType,
  State,
  void,
  Action<string>
>;

export type AppDispatch = ThunkDispatch<State, void, Action<string>>;
