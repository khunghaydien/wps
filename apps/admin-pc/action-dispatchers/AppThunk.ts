import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

// import { State } from '../modules';

export type AppAction<ReturnType = void> = ThunkAction<
  ReturnType,
  any,
  void,
  Action<string>
>;

export type AppDispatch = ThunkDispatch<any, void, Action<string>>;
