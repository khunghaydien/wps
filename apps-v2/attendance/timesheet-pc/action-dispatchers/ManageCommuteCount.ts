import {
  catchApiError,
  loadingEnd,
  loadingStart,
} from '../../../commons/actions/app';

import repository from '@attendance/repositories/ManageCommuteCountRepository';

import { CommuteCount } from '@attendance/domain/models/CommuteCount';

import { actions } from '../modules/entities/timesheet';

import { AppDispatch } from './AppThunk';

export const update =
  ({
    commuteCount,
    targetDate,
    employeeId,
  }: {
    commuteCount: CommuteCount;
    targetDate: string;
    employeeId?: string;
  }) =>
  async (dispatch: AppDispatch) => {
    try {
      dispatch(loadingStart());
      await repository.update({
        commuteCount,
        targetDate,
        employeeId,
      });
      dispatch(actions.updateCommuteCount(targetDate, commuteCount));
    } catch (err) {
      dispatch(catchApiError(err));
    } finally {
      dispatch(loadingEnd());
    }
  };
