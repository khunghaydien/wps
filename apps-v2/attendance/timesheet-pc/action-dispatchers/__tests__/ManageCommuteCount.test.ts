import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import RemoteError from '../../../../commons/errors/RemoteError';

import ApiMock from '../../../../../__tests__/mocks/ApiMock';
import { AppDispatch } from '../AppThunk';
import { update } from '../ManageCommuteCount';

const mockStore = configureMockStore([thunk]);

beforeEach(() => {
  ApiMock.reset();
  ApiMock.invoke.mockReset();
});
describe('update()', () => {
  test('should be update.', async () => {
    // Arrange
    ApiMock.invoke.mockResolvedValueOnce();
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;

    // Act
    await dispatch(
      update({
        commuteCount: {
          forwardCount: 0,
          backwardCount: 1,
        },
        targetDate: '2020-01-01',
        employeeId: 'employeeId',
      })
    );

    // Assert
    // API should be called.
    expect(ApiMock.invoke).toBeCalledWith({
      path: '/att/daily-commute-count/save',
      param: {
        empId: 'employeeId',
        targetDate: '2020-01-01',
        commuteForwardCount: 0,
        commuteBackwardCount: 1,
      },
    });
    // Actions should be called as correct order.
    expect(store.getActions()).toMatchSnapshot();
  });

  test('should be update if employeeId is null.', async () => {
    // Arrange
    ApiMock.invoke.mockResolvedValueOnce();
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;

    // Act
    await dispatch(
      update({
        commuteCount: {
          forwardCount: 0,
          backwardCount: 1,
        },
        targetDate: '2020-01-01',
      })
    );

    // Assert
    // API should be called.
    expect(ApiMock.invoke).toBeCalledWith({
      path: '/att/daily-commute-count/save',
      param: {
        empId: undefined,
        targetDate: '2020-01-01',
        commuteForwardCount: 0,
        commuteBackwardCount: 1,
      },
    });
    // Actions should be called as correct order.
    expect(store.getActions()).toMatchSnapshot();
  });

  test('should be called catchApiError() if API returns a error.', async () => {
    // Arrange
    ApiMock.invoke.mockRejectedValueOnce(
      new RemoteError({
        isSuccess: false,
        error: {
          errorCode: 'test',
          message: 'error test',
          stackTrace: 'stackTrace',
          groupCode: 1,
        },
      })
    );
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;

    // Act
    await dispatch(
      update({
        commuteCount: {
          forwardCount: 0,
          backwardCount: 1,
        },
        targetDate: '2020-01-01',
        employeeId: 'employeeId',
      })
    );

    // Assert
    // API should be called.
    expect(ApiMock.invoke).toBeCalledWith({
      path: '/att/daily-commute-count/save',
      param: {
        empId: 'employeeId',
        targetDate: '2020-01-01',
        commuteForwardCount: 0,
        commuteBackwardCount: 1,
      },
    });
    // Actions should be called as correct order.
    expect(store.getActions()).toMatchSnapshot();
  });
});
