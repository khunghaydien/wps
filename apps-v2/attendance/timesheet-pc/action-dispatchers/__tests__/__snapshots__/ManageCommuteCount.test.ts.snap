// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update() should be called catchApiError() if API returns a error. 1`] = `
Array [
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Object {
      "errorCode": "test",
      "isContinuable": true,
      "isFunctionCantUseError": false,
      "message": "APIエラー - error test",
      "problem": "error test",
      "stackTrace": "stackTrace",
      "type": "APIエラー",
    },
    "type": "CATCH_API_ERROR",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`update() should be update if employeeId is null. 1`] = `
Array [
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Object {
      "commuteCount": Object {
        "backwardCount": 1,
        "forwardCount": 0,
      },
      "targetDate": "2020-01-01",
    },
    "type": "TIMESHEET-PC/ENTITIES/TIMESHEET/UPDATE_COMMUTE_COUNT",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`update() should be update. 1`] = `
Array [
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Object {
      "commuteCount": Object {
        "backwardCount": 1,
        "forwardCount": 0,
      },
      "targetDate": "2020-01-01",
    },
    "type": "TIMESHEET-PC/ENTITIES/TIMESHEET/UPDATE_COMMUTE_COUNT",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;
